app/src/main/java/org/kiwix/kiwixmobile/zim_manager/library_view/LibraryFragment.java

```diff
@@ -163,10 +167,60 @@ public void showBooks(LinkedList<Book> books) {
     } else {
       libraryAdapter.getFilter().filter("", i -> stopScanningContent());
     }
-    libraryAdapter.notifyDataSetChanged();
+    onRefreshDisableListClick();
+    noRefreshQueryChange();
+    notifyDatachange();
     libraryList.setOnItemClickListener(this);
   }
 
+  public  void notifyDatachange(){
+    // we will notify the update every 0.3 second as the listitem position and content changes
+    final Handler handler = new Handler();
+    handler.postDelayed( new Runnable() {
+
+      @Override
+      public void run() {
+        activity.runOnUiThread(() -> libraryAdapter.notifyDataSetChanged());
+        handler.postDelayed( this, 300 );
+      }
+    }, 300 );
+  }
+  public void onRefreshDisableListClick(){
+    //  we will disable list while refreshing to prevent it from IndexOutofBoundException & IllegalException (notifyDataSetChanged())
+    if(swipeRefreshLayout.isRefreshing()){
+      libraryList.setEnabled(true);
+    }
+    else{
+      libraryList.setEnabled(false);
+    }
+  }
+
+  public void toastOnRefresh(){
+    // we will toast whenever user refreshes the list
+    if(swipeRefreshLayout.isRefreshing()){
+      Toast.makeText(getContext(),"Please wait while Refreshing",Toast.LENGTH_LONG).show();
+    }
+  }
+  public void noRefreshQueryChange(){
+    // on every querytextchange we will disable list action for 1sec to avoid sudden Exception IllegalException( notifyDataSetChanged())
+    Timer timer = new Timer();
+    libraryList.setEnabled(false);
+    if(activity.searchView.getQuery().length()>0){
+    timer.schedule(new TimerTask() {
```

Comment:
This could be simplified using Handler.

Reply:
sir its working quite well from preventing the app crash. 
Do you still want me to choose handler?
then ill be using it with my next commit
Thanks again :)